lib/chef/provider/package.rb

```diff
@@ -414,19 +433,19 @@ def target_version_array
             each_package do |package_name, new_version, current_version, candidate_version|
               case action
               when :upgrade
-                if version_equals?(current_version, new_version)
-                  # this is an odd use case
-                  logger.trace("#{new_resource} #{package_name} #{new_version} is already installed -- you are equality pinning with an :upgrade action, this may be deprecated in the future")
-                  target_version_array.push(nil)
-                elsif version_equals?(current_version, candidate_version)
-                  logger.trace("#{new_resource} #{package_name} #{candidate_version} is already installed")
+                if current_version.nil?
+                  logger.trace("#{new_resource} has no existing installed version. Installing install #{candidate_version}")
+                  target_version_array.push(candidate_version)
+                elsif version_equals?(current_version, new_version)
+                  # this is a short-circuit to avoid needing to (expensively) query the candidate_version which must come later
+                  logger.trace("#{new_resource} #{package_name} #{new_version} is already installed")
```

Comment:
I would put the warning back. pinning in :upgrade is weird and we still may kill it

Reply:
even if you delete this line because of the way that the candidate and installed versions are both filtered by the new_resource.version and/or the version constraints in the package_name it still works.  that line (at least for the DNF package provider now and soon the yum provider) is just a performance optimization now if you equality match (to avoid querying the candidate_version, which can be expensive in some of the subclasses).

Reply:
(and we haven't deleted in that in the years that comment has been sitting there, i'm pretty confident at this point that we never will)

Reply:
yeah I was totally wrong. This is a valid thing to do now. You might want to upgrade to whatever the latest release of 1.2 is